src/file/reader.rs

```diff
@@ -39,51 +39,49 @@ use schema::types::{self, SchemaDescriptor, Type as SchemaType};
 use thrift::protocol::TCompactInputProtocol;
 use util::io::FileSource;
 use util::memory::ByteBufferPtr;
+use std::io::Cursor;
 
 // ----------------------------------------------------------------------
 // APIs for file & row group readers
 
 /// Parquet file reader API. With this, user can get metadata information about the
 /// Parquet file, can get reader for each row group, and access record iterator.
-pub trait FileReader {
+pub trait FileReader<P: PageReader, R: RowGroupReader<P>> {
   /// Get metadata information about this file.
   fn metadata(&self) -> ParquetMetaDataPtr;
 
   /// Get the total number of row groups for this file.
   fn num_row_groups(&self) -> usize;
 
   /// Get the `i`th row group reader. Note this doesn't do bound check.
-  fn get_row_group(&self, i: usize) -> Result<Box<RowGroupReader>>;
+  fn get_row_group(&self, i: usize) -> Result<R>;
 
   /// Get full iterator of `Row`s from a file (over all row groups).
   ///
   /// Iterator will automatically load the next row group to advance.
   ///
   /// Projected schema can be a subset of or equal to the file schema, when it is None,
   /// full file schema is assumed.
-  fn get_row_iter(&self, projection: Option<SchemaType>) -> Result<RowIter>;
+  fn get_row_iter(&self, projection: Option<SchemaType>) -> Result<RowIter<P, R>>;
 }
 
 /// Parquet row group reader API. With this, user can get metadata information about the
 /// row group, as well as readers for each individual column chunk.
-pub trait RowGroupReader {
+pub trait RowGroupReader<P: PageReader>: Sized {
   /// Get metadata information about this row group.
   fn metadata(&self) -> RowGroupMetaDataPtr;
 
   /// Get the total number of column chunks in this row group.
   fn num_columns(&self) -> usize;
 
-  /// Get page reader for the `i`th column chunk.
-  fn get_column_page_reader(&self, i: usize) -> Result<Box<PageReader>>;
```

Comment:
I saw that this was only called by `get_column_reader`. It's hard to maintain this with the new generic interface. If this is important (as the trait is public), I can try to bring back this method.